ratpack-bearer-auth/src/main/java/st/ratpack/auth/CachingTokenValidator.java

```diff
@@ -4,44 +4,88 @@
 	private static Logger logger = LoggerFactory.getLogger(CachingTokenValidator.class);
 
-	public CachingTokenValidator(TokenValidator upstreamValidator) {
+	public CachingTokenValidator(TokenValidator upstreamValidator,
+	                             long timoutMinutes, TimeUnit timeUnit,
+	                             boolean refresh,
+	                             ExecController execController) {
 		this.upstreamValidator = upstreamValidator;
+		this.execController = execController;
 
-		cache = Caffeine.<String, Promise<Optional<OAuthToken>>>newBuilder()
-			.maximumSize(10000L)
-			.expireAfterWrite(5L, TimeUnit.MINUTES)
-			//			.executor(Execution.current().getEventLoop())  Don't do this it makes Ratpack hang.
-			.build(this::loadCache);
+		if (refresh) {
+			cache = Caffeine.<String, Promise<Optional<OAuthToken>>>newBuilder()
+					.maximumSize(10000L)
+					.refreshAfterWrite(timoutMinutes, timeUnit)
+					.build(this::loadCacheAsync);
+		} else {
+			cache = Caffeine.<String, Promise<Optional<OAuthToken>>>newBuilder()
+					.maximumSize(10000L)
+					.expireAfterWrite(timoutMinutes, timeUnit)
+					.build(this::loadCache);
+		}
 	}
 
 	@Override
 	public Promise<Optional<OAuthToken>> validate(String token) {
 		return cache.get(token);
 	}
 
+	private Promise<Optional<OAuthToken>> loadCacheAsync(String token) {
+		AtomicReference<Promise<Optional<OAuthToken>>> reference = new AtomicReference<>();
+		CountDownLatch latch = new CountDownLatch(1);
+		execController.fork()
+				.start(execution -> {
+					reference.set(loadCache(token));
+					latch.countDown();
+				});
+		try {
+			latch.await(1, TimeUnit.MINUTES);
+		} catch (InterruptedException ex) {
+			logger.error("cache load interrupt", ex);
+		}
+
+		Promise<Optional<OAuthToken>> result = reference.get();
+		logger.trace("PUTTING: {} {}", token, result);
+		return result;
+	}
+
 	private Promise<Optional<OAuthToken>> loadCache(String token) {
 		logger.trace("Cache MISS: {}", token);
 		//Make sure we only call the validate on the upstream once
 		Promise<Optional<OAuthToken>>
-				promiseOAuthToken = upstreamValidator.validate(token).cache();
-
-		promiseOAuthToken
-				.onError(e -> cache.invalidate(token))
-				.map(o -> o.orElse(null))
-				.onNull(() -> cache.invalidate(token))
-				.then(o -> logger.trace("PUTTING: {}", o));
+				promiseOAuthToken = upstreamValidator.validate(token)
+				.next(o -> {
+					// this should just get invoked once
+					if (!o.isPresent()) {
+						// evict and let another request try
+						logger.trace("EVICT NOT PRESENT: {}", o);
+						cache.invalidate(token);
+					} else {
+						logger.trace("PROMISE RESOLVED: {}", token);
+					}
+				})
+				.onError(e -> {
+					logger.warn("error validating token", e);
+					cache.invalidate(token); // evict and let another request try
+				})
 
+				// Cache the promise execution the first time it is used so when it is read
+				// the wrapped validator does not get called again
+				.cache();
```

Comment:
Should this be last?

Reply:
I switched to mapError maybe it should just do the evict and then re-throw (`Promise.error(e)`)

Reply:
That cache just means we won't re evaluate the promise just use the value retrieved the first time

Reply:
right it only needs to do that stuff once